File: src/sources/nih.ts
import Parser from 'rss-parser';
import { BaseSource, RawArticle } from './base';
import { logger } from '@/lib/logger';

const parser = new Parser({ timeout: 15000, maxRedirects: 5 });

// NIH 공식 RSS는 Cloudflare가 봇 차단 → ScienceDaily로 대체 (NIH 연구 집중 보도)
const FEEDS = [
  'https://www.sciencedaily.com/rss/health_medicine.xml',
  'https://www.sciencedaily.com/rss/top/health.xml',
];

export class NIHSource extends BaseSource {
  readonly name = 'ScienceDaily';

  async fetch(): Promise<RawArticle[]> {
    const articles: RawArticle[] = [];
    for (const feedUrl of FEEDS) {
      try {
        const feed = await parser.parseURL(feedUrl);
        for (const item of feed.items) {
          if (!item.title || !item.link) continue;
          articles.push({
            source: this.name,
            title: this.stripHtml(item.title),
            url: item.link,
            published_at: this.parseDate(item.pubDate ?? item.isoDate),
            summary: item.contentSnippet ?? '',
            language: this.language,
          });
        }
      } catch (e) {
        logger.warn(`[ScienceDaily] 피드 오류 ${feedUrl}: ${e}`);
      }
    }
    return articles;
  }
}

File: src/sources/nih.py
import feedparser
from .base import BaseSource, RawArticle


class NIHSource(BaseSource):
    name = "NIH"
    language = "en"
    FEEDS = [
        "https://www.nih.gov/news-events/news-releases/feed",
        "https://www.niaid.nih.gov/news-events/news-releases/feed",  # NIAID (감염병)
    ]

    async def fetch(self) -> list[RawArticle]:
        articles = []
        for url in self.FEEDS:
            try:
                feed = feedparser.parse(url)
                for entry in feed.entries:
                    articles.append(RawArticle(
                        source=self.name,
                        title=entry.get("title", "").strip(),
                        url=entry.get("link", ""),
                        published_at=self._safe_date(entry.get("published")),
                        summary=entry.get("summary", ""),
                        language=self.language,
                    ))
            except Exception as e:
                from src.utils.logger import logger
                logger.warning(f"[NIH] 피드 오류 {url}: {e}")
        return articles
